### ISSUE TITLE
Models Now Transformed with Y=180 Degree Rotation

### ISSUE THREAD

**Issue by KermMartian** [KhronosGroup/UnityGLTF#319]
Pretty much what it says on the tin. As of a version from around March 2018, models were properly loaded with the transform offset from the loaded GLTF's (0, 0, 0) point. ~~Now, it appears that models are transformed from the center of the model, which is very challenging for precise positioning. Is this intentional behavior, and if so, is there a way to work around it?~~

I'm afraid I had identified the symptom of this problem incorrectly. What appeared to be an incorrect offset was, in fact, an extra transformation of 180 degrees around the Y axis (in the XZ plane). In other words, while the model's (0, 0, 0) point is in the correct place, the model extends along the negative X and negative Z axis from that point, instead of the positive X and positive Z axis. I have temporarily worked around this by setting `obj.transform.eulerAngles = new Vector3(0f, 180f, 0f);` for each new `GLTFComponent` that I create, but that feels like a fragile solution indeed. I note that `UnityGLTF.Extensions.SchemaExtensions` has some comments that discuss GLTF handedness versus Unity handedness; might this be the culprit? I don't see a robust way to override this behavior. TIA.

**Comment by blgrossMS** [comment]
Whats an example model you noticed this behavior with?

**Comment by blgrossMS** [comment]
See the conversation in #257, does this address the coordinate transform issue you are seeing?

**Comment by otri** [comment]
I can confirm, coordinate spaces don't match export from Blender into Unity.  Also a direct comparison of FBX import vs GLB import, shows the GLB gets a Y=180 degree rotation.  Model contains identity transform, no rotation, scale, or translation applied.  Strictly handedness conversion gets the coordinate spaces wrong.  I think people expect Up and Forward orientation the scene space to be respected here, so a GLB scene is loaded identically with respect to other model imports.

FBX and GLB imports with correct relative orientation in Apple's Reality Converter and in Blender.  So it really appears to be an issue of UnityGLTF and glTFast import convention malpractice.

**Comment by hybridherbst** [comment]
Can you check from the dev branch please?

**Comment by XtroTheArctic** [comment]
I know this is a very old post but... I tried UnityGLTF last night for the first time and I experienced the same 180 Y rotation issue too but it turned out to be a human error in Blender object Transform; not a bug in UnityGLTF. The latest version of UnityGLTF as of today doesn't have this problem so this issue ticket may need to be closed.

**Comment by TRMMax** [comment]
Unlike @XtroTheArctic noted, it seems to still be an issue. Export a fresh cube with an identity transform but with its mesh deformed such that you can clearly see its orientation. Using +Y Up (obviously, otherwise the model is completely rotated wrongly since blenders axis are different), it will be imported into Unity with its mesh rotated 180 degrees (but not its transform! that is also an identity transform). This means that to get it identical in Blender you have to change the Unity transform to have a rotation of 180 degrees around the Y axis.

**Comment by XtroTheArctic** [comment]
Since my comment above, I have been using UnityGLTF extensively and I love it. I haven't had any orientation issue while exporting my models with identity transform from Blender to Unity.

I am using Blender 4.4.3

If you can reproduce the same problem with the cube test using that version, and record a video of it, I can try to repro the same thing on my side to make sure the problem actually exists. I'd be happy to help if I could.

**Comment by TRMMax** [comment]
I've have been able to reproduce it in Unity 6000.0.38f1, glTFast 6.12.1 (also UnityGLTF 2.16.1) and Blender 4.4.3
All settings are default (+Y Up is now default apparently) and its both a new Blender and Unity project.

https://github.com/user-attachments/assets/b8dbb4f5-a24a-491d-b549-d24b1412fcc8

Let me know if this is clear enough :)

**Comment by XtroTheArctic** [comment]
This is correct. In Unity, +Z (blue arrow) is the front of the object. And in Blender -Y (opposite side of green arrow) is the front of the object. Your export/import is working correctly.

If you press Numpad 1 in Blender, you will see the model from its front and this matches the front of the model in Unity shown as blue arrow.

**Comment by otri** [comment]
Here's the little test set-up in Blender v4.3.2 and Unity 6000.0.26f1.

[2025-06-29 Forward Check.zip](https://github.com/user-attachments/files/20506430/2025-06-29.Forward.Check.zip)

**Comment by hybridherbst** [comment]
This is a very fun topic and one of many discussions :)

First of all, I can ensure you that UnityGLTF does the "right thing" both when importing, and when exporting. Doing the right thing means that for some nodes, we need to introduce "flip transforms" on import or export, because the definitions of "forward" do not align between engines, nodes, cameras, lights, default views, texture coordinates and so on, things could be animated, … – in any combination.

Some notes:

Unity's definition of "forward" aligns with an audience of game developers making third-person games. "Forward" in Unity means: the camera looks along +Z, and by default sits at Z=-10. Now, that means, when you add a cube in Unity, by default you look – at its back! It's +Z axis points _away_ from the default camera. So, when you want to make a "3D Model Viewer", the right thing to do as very first action is to move Unity's default camera to the _other side_ and rotate it so it looks at the front of models you load into the engine. If you don't do this, you'll be stuck with 180° flips for all your model imports.

Now, in Blender, the "forward axis" for cameras is +Y, and the "forward axis" for models is … -Y! You can see that when you create the default Suzanne head; and then choose "View > Viewpoint > Front". And the default camera (in contrast to what it does in Unity) is set up for the typical "model viewing" case, which means it by default, without orientation, looks at the front of the model.

So not only do coordinate systems differ between all three of glTF, Unity, and Blender, but also what "forward" means for a camera is different. Unfortunately, due to the way data is stored in glTF, we can't just "rotate the camera" on import; that would break animations of the camera, which expect the data to be "in glTF space". Thus, the inbetween node is needed.

**Comment by otri** [comment]
That makes so much more sense now.  Oh goodness, what confounding rules and relationships. @hybridherbst thank you for kindly explaining it in this way.
